File: samples/ObjC Automatic Sample/Sample/CustomActionVC.m
/*
 * Copyright © 2019, 2019 Acoustic, L.P. All rights reserved.
 *
 * NOTICE: This file contains material that is confidential and proprietary to
 * Acoustic, L.P. and/or other developers. No license is granted under any intellectual or
 * industrial property rights of Acoustic, L.P. except as may be provided in an agreement with
 * Acoustic, L.P. Any unauthorized copying or distribution of content from this file is
 * prohibited.
 */

#import "CustomActionVC.h"
#import "UIColor+Sample.h"

#if __has_feature(modules)
@import AcousticMobilePush;
#else
#import <AcousticMobilePush/AcousticMobilePush.h>
#endif

NSString * customActionUserDefaultKey = @"customActions";
NSString * typeUserDefaultKey = @"customActionType";
NSString * valueUserDefaultKey = @"customActionValue";

@interface CustomActionVC ()
@property NSMutableArray * registeredTypes;
@end

@implementation CustomActionVC

-(void)viewWillAppear:(BOOL)animated {
    self.registeredTypes = [NSMutableArray array];
    [super viewWillAppear:animated];
    [self updateTheme];
    
    for (NSString * type in [NSUserDefaults.standardUserDefaults stringArrayForKey: customActionUserDefaultKey]) {
        [MCEActionRegistry.sharedInstance registerTarget:self withSelector:@selector(receiveCustomAction:) forAction: type];
    }
    
    self.typeField.text = [NSUserDefaults.standardUserDefaults stringForKey: typeUserDefaultKey];
    self.valueField.text = [NSUserDefaults.standardUserDefaults stringForKey: valueUserDefaultKey];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    for (NSString * type in self.registeredTypes) {
        [MCEActionRegistry.sharedInstance unregisterAction: type];
    }
    [self.registeredTypes removeAllObjects];
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
    self.keyboardDoneButtonView = [[UIToolbar alloc] init];
    self.keyboardDoneButtonView.barStyle = UIBarStyleDefault;
    self.keyboardDoneButtonView.translucent = YES;
    self.keyboardDoneButtonView.tintColor = nil;
    
    UIBarButtonItem * doneButton = [[UIBarButtonItem alloc] initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(doneClicked:)];
    doneButton.accessibilityIdentifier = @"doneButton";
    self.keyboardDoneButtonView.items = @[ doneButton ];
    
    [NSNotificationCenter.defaultCenter addObserverForName:MCECustomPushNotYetRegistered object:nil queue: NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        NSDictionary * action = note.userInfo[@"action"];
        self.statusLabel.textColor = UIColor.warningColor;
        self.statusLabel.text = [NSString stringWithFormat: @"Previously Registered Custom Action Received: %@", action];
    }];

    [NSNotificationCenter.defaultCenter addObserverForName:MCECustomPushNotRegistered object:nil queue: NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        NSDictionary * action = note.userInfo[@"action"];
        self.statusLabel.textColor = UIColor.failureColor;
        self.statusLabel.text = [NSString stringWithFormat: @"Unregistered Custom Action Received: %@", action];
    }];

    [NSNotificationCenter.defaultCenter addObserverForName:UIKeyboardWillChangeFrameNotification object:nil queue: NSOperationQueue.mainQueue usingBlock:^(NSNotification * _Nonnull note) {
        CGRect endFrame = [note.userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
        NSTimeInterval duration = [note.userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
        UIViewAnimationOptions options = [note.userInfo[UIKeyboardAnimationCurveUserInfoKey] integerValue];
        if(endFrame.origin.y >= UIScreen.mainScreen.bounds.size.height) {
            self.keyboardHeightLayoutConstraint.constant = 0.0;
        } else {
            self.keyboardHeightLayoutConstraint.constant = endFrame.size.height;
        }
        
        [UIView animateWithDuration:duration delay:0 options:options animations:^{
            [self.view layoutIfNeeded];
        } completion:^(BOOL finished) {
            
        }];
    }];
}

-(void) doneClicked: (id)sender {
    [self.typeField resignFirstResponder];
    [self.valueField resignFirstResponder];
}

- (void)textFieldDidBeginEditing:(UITextField *)textField {
    textField.inputAccessoryView = self.keyboardDoneButtonView;
    [self.keyboardDoneButtonView sizeToFit];
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    [NSUserDefaults.standardUserDefaults setObject: self.typeField.text forKey: typeUserDefaultKey];
    [NSUserDefaults.standardUserDefaults setObject: self.valueField.text forKey: valueUserDefaultKey];
}

// This method simulates how custom actions receive push actions
-(void)receiveCustomAction:(NSDictionary *) action {
    if(![NSThread isMainThread]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self receiveCustomAction: action];
        });
        return;
    }

    if([self.registeredTypes containsObject: action[@"type"]]) {
        self.statusLabel.textColor = UIColor.successColor;
        self.statusLabel.text = [NSString stringWithFormat: @"Received Custom Action: %@", action];
    } else {
        self.statusLabel.textColor = UIColor.warningColor;
        self.statusLabel.text = [NSString stringWithFormat: @"Previously Registered Custom Action Received: %@", action];
    }
}

// This method simulates a custom action registering to receive push actions
- (IBAction)registerCustomAction:(id)sender {
    NSMutableSet * customActions = [NSMutableSet setWithArray: [NSUserDefaults.standardUserDefaults stringArrayForKey:customActionUserDefaultKey]];
    [customActions addObject:self.typeField.text];
    [NSUserDefaults.standardUserDefaults setObject: [customActions allObjects] forKey: customActionUserDefaultKey];
    
    [self.registeredTypes addObject: self.typeField.text];
    self.statusLabel.textColor = UIColor.successColor;
    self.statusLabel.text = [NSString stringWithFormat: @"Registering Custom Action: %@", self.typeField.text];
    [MCEActionRegistry.sharedInstance registerTarget:self withSelector:@selector(receiveCustomAction:) forAction:self.typeField.text];
}

// This method simulates a user clicking on a push message with a custom action
- (IBAction)sendCustomAction:(id)sender {
    NSDictionary * action = @{@"type": self.typeField.text, @"value": self.valueField.text};
    NSDictionary * payload = @{@"notification-action": action};
    self.statusLabel.textColor = UIColor.successColor;
    self.statusLabel.text = [NSString stringWithFormat: @"Sending Custom Action: %@", action];
    [MCEActionRegistry.sharedInstance performAction:action forPayload:payload source:@"internal" attributes:nil userText:nil];
}

// This method shows how to unregister a custom action
- (IBAction)unregisterCustomAction:(id)sender {
    self.statusLabel.textColor = UIColor.successColor;
    self.statusLabel.text = [NSString stringWithFormat: @"Unregistered Action: %@", self.typeField.text];
    [MCEActionRegistry.sharedInstance unregisterAction:self.typeField.text];
}

-(void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    [self updateTheme];
}

-(void)updateTheme {
    if(![NSThread isMainThread]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self updateTheme];
        });
        return;
    }
    self.typeField.textColor = UIColor.foregroundColor;
    self.valueField.textColor = UIColor.foregroundColor;
}

@end

File: samples/Swift Sample/Shared Sample/CustomActionVC.swift
/*
 * Copyright © 2019, 2019 Acoustic, L.P. All rights reserved.
 *
 * NOTICE: This file contains material that is confidential and proprietary to
 * Acoustic, L.P. and/or other developers. No license is granted under any intellectual or
 * industrial property rights of Acoustic, L.P. except as may be provided in an agreement with
 * Acoustic, L.P. Any unauthorized copying or distribution of content from this file is
 * prohibited.
 */

import UIKit

class CustomActionVC: UIViewController, MCEActionProtocol {
    var _interfaceState: Data? = nil
    
    @IBOutlet weak var typeField: UITextField!
    @IBOutlet weak var valueField: UITextField!
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var keyboardHeightLayoutConstraint: NSLayoutConstraint?
    
    var registeredTypes = [String]()
    let keyboardToolbar = UIToolbar()
    
    let colors = (
        error: UIColor(red: 0.5, green: 0, blue: 0, alpha: 1),
        success: UIColor(red: 0, green: 0.5, blue: 0, alpha: 1),
        warning: UIColor(red: 0.574, green: 0.566, blue: 0, alpha: 1)
    )
        
    // This method simulates a custom action registering to receive push actions
    @IBAction func registerCustomAction(_ sender: Any) {
        guard let string = typeField.text else {
            return
        }
        
        registerCustomAction(string: string)
    }
    
    func registerCustomAction(string: String) {
        registeredTypes.append(string)
        var customActions = Set( UserDefaults.standard.stringArray(forKey: "customActions") ?? [String]() )
        customActions.insert(string)
        UserDefaults.standard.set(Array(customActions), forKey: "customActions")
        
        statusLabel.text = "Registering Custom Action: \(string)"
        statusLabel.textColor = colors.success
        MCEActionRegistry.shared.registerTarget(self, with: #selector(self.receiveCustomAction(action:)), forAction: string)
    }
    
    // This method shows how to unregister a custom action
    @IBAction func unregisterCustomAction(_ sender: Any) {
        guard let type = typeField.text else {
            return
        }
        registeredTypes.removeAll(where: { (registeredType) -> Bool in
            return registeredType == type
        })
        statusLabel.textColor = colors.success
        statusLabel.text = "Unregistered Action: \(type)"
        MCEActionRegistry.shared.unregisterAction(type)
    }
    
    // This method simulates a user clicking on a push message with a custom action
    @IBAction func sendCustomAction(_ sender: Any) {
        guard let type = typeField.text, let value = valueField.text else {
            return
        }
        let action = ["type": type, "value": value]
        let payload = ["notification-action": action]
        statusLabel.textColor = colors.success
        statusLabel.text = "Sending Custom Action: \(action)"
        MCEActionRegistry.shared.performAction(action, forPayload: payload, source: "internal", attributes: nil, userText: nil)
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        keyboardToolbar.barStyle = .default
        keyboardToolbar.isTranslucent = true
        keyboardToolbar.tintColor = nil
        
        let doneButton = UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(doneClicked))
        doneButton.accessibilityIdentifier = "doneButton"
        keyboardToolbar.items = [doneButton]
        
        NotificationCenter.default.addObserver(forName: MCENotificationName.customPushNotYetRegistered.rawValue, object: nil, queue: .main) { (note) in
            guard let action = note.userInfo?["action"] as? [String:Any] else {
                return
            }
            self.statusLabel.textColor = self.colors.warning
            self.statusLabel.text = "Previously Registered Custom Action Received: \(action)"
        }
        
        NotificationCenter.default.addObserver(forName: MCENotificationName.customPushNotRegistered.rawValue, object: nil, queue: .main) { (note) in
            guard let action = note.userInfo?["action"] as? [String:Any] else {
                return
            }
            self.statusLabel.textColor = self.colors.error
            self.statusLabel.text = "Unregistered Custom Action Received: \(action)"
        }
        
        NotificationCenter.default.addObserver(forName: UIWindow.keyboardWillChangeFrameNotification, object: nil, queue: .main) { (note) in
            if let keyboardHeightLayoutConstraint = self.keyboardHeightLayoutConstraint, let userInfo = note.userInfo,
                let frame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect,
                let duration = userInfo[UIResponder.keyboardAnimationDurationUserInfoKey] as? TimeInterval,
                let optionsInteger = userInfo[UIResponder.keyboardAnimationCurveUserInfoKey] as? UInt {
                let options = UIView.AnimationOptions(rawValue: optionsInteger)
                
                if frame.origin.y >= UIScreen.main.bounds.size.height {
                    keyboardHeightLayoutConstraint.constant = 0
                } else {
                    keyboardHeightLayoutConstraint.constant = frame.size.height
                }
                
                UIView.animate(withDuration: duration, delay: 0, options: options, animations: {
                    self.view.layoutIfNeeded()
                })
            }
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        for type in registeredTypes {
            MCEActionRegistry.shared.unregisterAction(type)
        }
        registeredTypes.removeAll()
    }
        
    override func viewDidLoad() {
        super.viewDidLoad()
        updateTheme()
    }
    
    @objc func doneClicked() {
        typeField.resignFirstResponder()
        valueField.resignFirstResponder()
    }
    

    // This method simulates how custom actions receive push actions
    @objc func receiveCustomAction(action: NSDictionary) {
        DispatchQueue.main.async {
            self.statusLabel.text = "Received Custom Action: \(action)"
            self.statusLabel.textColor = self.colors.success
        }
    }
    
    func updateTheme() {
        typeField.textColor = .foreground
        valueField.textColor = .foreground
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        updateTheme()
    }
}

extension CustomActionVC: UITextFieldDelegate {
    func textFieldDidEndEditing(_ textField: UITextField, reason: UITextField.DidEndEditingReason) {
        guard reason == .committed else {
            return
        }
        UserDefaults.standard.set(["interfaceState": interfaceState], forKey: String(describing: type(of: self)))
    }
    
    func textFieldDidBeginEditing(_ textField: UITextField) {
        textField.inputAccessoryView = keyboardToolbar
        keyboardToolbar.sizeToFit()
    }
}

// Generic State Restoration
extension CustomActionVC: RestorableVC {
    struct InterfaceState: Codable {
        let typeField: String
        let valueField: String
        let statusText: String
        let statusColor: Data
        var editingField: TextField? = nil
        var editingValue: String? = nil
        var registeredTypes: [String]
    }

    enum TextField: String, Codable {
        case typeField
        case valueField
    }
    
    var interfaceState: Data? {
        get {
            var interfaceState = InterfaceState(
                typeField: typeField.text ?? "",
                valueField: valueField.text ?? "",
                statusText: self.statusLabel.text ?? "No status yet",
                statusColor: self.statusLabel.textColor.data,
                registeredTypes: registeredTypes
            )

            if typeField.isFirstResponder {
                interfaceState.editingField = .typeField
                interfaceState.editingValue = typeField.text
            } else if valueField.isFirstResponder {
                interfaceState.editingField = .valueField
                interfaceState.editingValue = valueField.text
            }
                        
            do {
                return try PropertyListEncoder().encode(interfaceState)
            } catch {
                print("Could not encode interface state \(error)")
                return nil
            }
        } set {
            _interfaceState = newValue
        }
    }
    
    func restoreInterfaceStateIfAvailable() {
        guard let data = _interfaceState else {
            return
        }
        
        let interfaceState: InterfaceState
        do {
            interfaceState = try PropertyListDecoder().decode(InterfaceState.self, from: data)
        } catch {
            print("Could not decode interface state \(error)")
            return
        }

        for string in interfaceState.registeredTypes {
            registerCustomAction(string: string)
        }

        typeField.text = interfaceState.typeField
        valueField.text = interfaceState.valueField
        
        statusLabel.text = interfaceState.statusText
        statusLabel.textColor = UIColor.from(data: interfaceState.statusColor) ?? UIColor.disabled
        
        if let editingField = interfaceState.editingField, let value = interfaceState.editingValue {
            switch editingField {
            case .typeField:
                typeField.text = value
                typeField.becomeFirstResponder()
            case .valueField:
                valueField.text = value
                valueField.becomeFirstResponder()
            }
        }
        
        _interfaceState = nil
    }
    
}

// State restoration iOS ≤12
extension CustomActionVC {
    
    override func encodeRestorableState(with coder: NSCoder) {
        super.encodeRestorableState(with: coder)
        coder.encode(interfaceState, forKey: "interfaceState")
    }
    
    override func decodeRestorableState(with coder: NSCoder) {
        super.decodeRestorableState(with: coder)
        if let interfaceState = coder.decodeObject(forKey: "interfaceState") as? Data {
            self.interfaceState = interfaceState
        }
    }
}

// State Restoration and Multiple Window Support iOS ≥13
extension CustomActionVC {

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        restoreInterfaceStateIfAvailable()
    }
}
